bug: Numeric literal suffix with generic type ICEs compiler

# Bug Report

**Cairo version:**

`6a3c423a9`

**Current behavior:**

`cairo-compile` panics with an itertools `.zip_eq()` invariant violation when a numeric literal uses a corelib `pub enum` / `pub struct` type with required generic parameters as its suffix. The four affected type names are `Span`, `Option`, `Result`, and `PanicResult`. The crash exits without a clean diagnostic E-code:

```
thread 'main' panicked at zip_eq_impl.rs:50:17:
itertools: .zip_eq() reached end of one iterator before the other
```

An empirical sweep over 27 corelib type names used as numeric-literal suffixes (`42_<TypeName>`) at the pinned commit produces a 5-category typology:

- **Cat A — ICE (4 types):** `Span`, `Option`, `Result`, `PanicResult` — itertools `.zip_eq()` panic, no clean E-code. All are corelib `pub enum` / `pub struct` types with required generic parameters.
- **Cat B — clean compile (12 types):** `u8`, `u16`, `u32`, `u64`, `u128`, `u256`, `i8`, `i16`, `i32`, `i64`, `i128`, `felt252` — the legitimate numeric-literal-suffix types.
- **Cat C — E2008 graceful (8 types):** `Array`, `Box`, `Felt252Dict`, `Nullable`, `bool`, `ByteArray`, `Pedersen`, `EcPoint` — mostly `pub extern type`; matcher emits `'A numeric literal of type <T> cannot be created.'`
- **Cat D — E2006 not found (2 types):** `Bytes31`, `EnumeratedIter`.
- **Cat E — trait/non-type (4 names):** `Default`, `Drop`, `Copy` → E2011; `NonZero` → E2030.

The inconsistency: Cat-C generic types declared as `pub extern type` (e.g., `pub extern type Array<T>;`) take a branch in the suffix matcher that emits E2008, while Cat-A `pub enum` / `pub struct` types with required generics fall into the `.zip_eq()` path where a length mismatch (e.g., 1 expected vs. 0 supplied generic args for `Option<T>`) panics.

PR #9810 (commit `a459dfa0e`, April 2026, included in `6a3c423a9`) emits a diagnostic for non-type identifier suffixes instead of panicking, but it does not cover the Cat-A `pub enum` / `pub struct`-with-required-generics case, which still ICEs.

**Expected behavior:**

`cairo-compile` should emit a graceful diagnostic (e.g., the same `error[E2008]: A numeric literal of type <T> cannot be created.` already produced for Cat-C extern types) instead of panicking with an internal `.zip_eq()` invariant violation. Suffix matching for `pub enum` / `pub struct` corelib types with required generics should be consistent with the extern-type path.

**Steps to reproduce:**

1. Build / use `cairo-compile` at commit `6a3c423a9`.
2. Save the snippet below as `probe.cairo`.
3. Run `cairo-compile --single-file probe.cairo`.
4. Observe the `.zip_eq()` panic instead of a diagnostic.

```bash
$ echo 'fn main() { let _x = 42_Span; }' > probe.cairo
$ cairo-compile --single-file probe.cairo
thread 'main' panicked at zip_eq_impl.rs:50:17:
itertools: .zip_eq() reached end of one iterator before the other
```

For comparison, the Cat-C path emits a clean diagnostic:

```cairo
fn main() { let _x = 42_Array; }
```

```
error[E2008]: A numeric literal of type core::array::Array cannot be created.
 --> probe.cairo:X:Y
    let _x = 42_Array;
             ^^^^^^^^
```

**Related code:**

<!-- If you are able to illustrate the bug or feature request with an example, please provide it here. -->

All four Cat-A suffixes ICE the compiler:

```cairo
fn main() {
    let _a = 42_Span;         // ICE
    let _b = 42_Option;       // ICE
    let _c = 42_Result;       // ICE
    let _d = 42_PanicResult;  // ICE
}
```

The relevant corelib type definitions (each requires generic args):

```
pub enum Option<T> { Some(T), None }
pub enum Result<T, E> { Ok(T), Err(E) }
pub enum PanicResult<T> { Ok(T), Err: (Panic, Array<felt252>) }
pub struct Span<T> { snapshot: @Array<T> }
```

Contrast with the Cat-C extern types that handle the same situation gracefully:

```
pub extern type Array<T>;
pub extern type Box<T>;
pub extern type Nullable<T>;
pub extern type Felt252Dict<T>;
```

**Other information:**

- N/A

--- a/crates/cairo-lang-semantic/src/expr/compute.rs
+++ b/crates/cairo-lang-semantic/src/expr/compute.rs
@@ -55,7 +55,7 @@ use super::pattern::{
     PatternOtherwise, PatternTuple, PatternVariable, PatternWrappingInfo,
 };
 use crate::corelib::{
-    self, CorelibSemantic, core_binary_operator, core_bool_ty, core_unary_operator,
+    self, CorelibSemantic, LiteralError, core_binary_operator, core_bool_ty, core_unary_operator,
     false_literal_expr, get_usize_ty, never_ty, true_literal_expr, try_extract_box_inner_type,
     try_get_core_ty_by_name, unit_expr, unit_ty, unwrap_error_propagation_type, validate_literal,
 };
@@ -3682,7 +3682,14 @@ fn new_literal_expr<'db>(
         let ty = try_get_core_ty_by_name(ctx.db, ty_str, vec![])
             .map_err(|err| ctx.diagnostics.report(stable_ptr.untyped(), err))?;
         if let Err(err) = validate_literal(ctx.db, ty, &value) {
-            ctx.diagnostics.report(stable_ptr, SemanticDiagnosticKind::LiteralError(err));
+            let is_invalid_type = matches!(err, LiteralError::InvalidTypeForLiteral(_));
+            let diag =
+                ctx.diagnostics.report(stable_ptr, SemanticDiagnosticKind::LiteralError(err));
+            // The type is malformed (generic type with zero args); returning Ok would ICE
+            // downstream queries via GenericSubstitution::new zip_eq mismatch.
+            if is_invalid_type {
+                return Err(diag);
+            }
         }
         return Ok(ExprNumericLiteral { value, ty, stable_ptr });
     };

--- a/crates/cairo-lang-semantic/src/expr/test_data/literal
+++ b/crates/cairo-lang-semantic/src/expr/test_data/literal
@@ -218,3 +218,42 @@ error[E2315]: Mismatched types. The type `()` cannot be created from a numeric l
  --> lib.cairo:3:5
     x
     ^
+
+//! > ==========================================================================
+
+//! > Numeric literal suffix with generic type emits E2008 instead of ICE (#9975).
+
+//! > test_runner_name
+test_function_diagnostics(expect_diagnostics: true)
+
+//! > function_code
+fn foo() {
+    let _a = 42_Option;
+    let _b = 42_Result;
+    let _c = 42_PanicResult;
+    let _d = 42_Span;
+}
+
+//! > function_name
+foo
+
+//! > expected_diagnostics
+error[E2008]: A numeric literal of type core::option::Option cannot be created.
+ --> lib.cairo:2:14
+    let _a = 42_Option;
+             ^^^^^^^^^
+
+error[E2008]: A numeric literal of type core::result::Result cannot be created.
+ --> lib.cairo:3:14
+    let _b = 42_Result;
+             ^^^^^^^^^
+
+error[E2008]: A numeric literal of type core::panics::PanicResult cannot be created.
+ --> lib.cairo:4:14
+    let _c = 42_PanicResult;
+             ^^^^^^^^^^^^^^
+
+error[E2008]: A numeric literal of type core::array::Span cannot be created.
+ --> lib.cairo:5:14
+    let _d = 42_Span;
+             ^^^^^^^

